### PR TITLE
net-p2p/go-ipfs: new package

### PR DESCRIPTION
**Pull request by stkw0**
non binary version of go-ipfs-bin

Package-Manager: Portage-2.3.68, Repoman-2.3.16
Signed-off-by: David Roman <davidroman96@gmail.com>

### CODE REVIEW

File: net-p2p/go-ipfs/go-ipfs-0.4.21.ebuild [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	systemd_dounit "${FILESDIR}/ipfs.service"
+	newinitd "${FILESDIR}/ipfs.init" ipfs
+	newconfd "${FILESDIR}/ipfs.confd" ipfs
+
+	newbashcomp "${DISTDIR}/${P}.bash" "ipfs"
+	keepdir /var/log/ipfs
+}
+
+pkg_preinst() {
+	enewgroup ipfs
```
Comment (juippis) [inline net-p2p/go-ipfs/go-ipfs-0.4.21.ebuild]: (This is related to your other PR, just follow the instructions there, and lets update this ebuild when the proper group/user additions are done.)

**Comment by kakra** [comment]
Let's first get `go-ipfs-bin` in because #12960 will also add the needed migration to `acct-*`. We should then update the ebuild and update it to the latest IPFS version. I can then look at it. @stkw0 Maybe you want to invite me as a contributor to your branch?

**Comment by stkw0** [comment]
> Let's first get `go-ipfs-bin` in because #12960 will also add the needed migration to `acct-*`. We should then update the ebuild and update it to the latest IPFS version. I can then look at it. @stkw0 Maybe you want to invite me as a contributor to your branch?

Totally agree, I've just invited as a collaborator.

File: net-p2p/go-ipfs/files/ipfs.init [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#!/sbin/openrc-run
+# Copyright 1999-2018 Gentoo Foundation
```
Comment (juippis) [inline net-p2p/go-ipfs/files/ipfs.init]: Please update.

File: net-p2p/go-ipfs/go-ipfs-0.4.23.ebuild [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+inherit bash-completion-r1 go-module golang-build systemd
+
+DESCRIPTION="Main implementation of IPFS"
+HOMEPAGE="https://ipfs.io/"
+SRC_URI="https://dist.ipfs.io/go-ipfs/v${PV}/go-ipfs-source.tar.gz -> ${P}.tar.gz
+	https://raw.githubusercontent.com/ipfs/go-ipfs/v${PV}/misc/completion/ipfs-completion.bash -> ${P}.bash"
+EGO_PN="github.com/ipfs/go-ipfs"
+
+LICENSE="MIT"
```
Comment (juippis) [inline net-p2p/go-ipfs/go-ipfs-0.4.23.ebuild]: You'll need to add licenses of bundled deps as well. I think `go-license-detector` can identify them, but I haven't used it.

File: net-p2p/go-ipfs/go-ipfs-0.4.23.ebuild [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+src_test() {
+	go test ${EGO_PN}/cmd/ipfs/... ${EGO_PN}/cmd/ipfswatch/... || die
+}
+
+src_install() {
+	dobin ipfs
+	dobin ipfswatch
+
```
Comment (juippis) [inline net-p2p/go-ipfs/go-ipfs-0.4.23.ebuild]: `einstalldocs` here somewhere? Will it change anything?

**Comment by stkw0** [comment]
> ```
>  * QA Notice: systemd units using /etc/conf.d detected:
>  * /lib/systemd/system/ipfs.service:EnvironmentFile=-/etc/conf.d/ipfs
>  * /lib/systemd/system/ipfs@.service:EnvironmentFile=-/etc/conf.d/ipfs.%i
>  * See: https://wiki.gentoo.org/wiki/Project:Systemd/conf.d_files
> ```
> 
> Is this a problem? I'm not too familiar with systemd. Looking at the wiki page,
> https://wiki.gentoo.org/wiki/Project:Systemd/conf.d_files#Inconsistent_system_configuration_handling
> 
> I think you may want to fix it before starting to get bug reports about it.

I changed it to use plain environment variables. I think that is the correct way to do it although I'm not familiar with systemd neither. @kakra I would appreciate if you can take a look just in case there is a better approach to this or I did something wrong.

**Comment by kakra** [comment]
> I changed it to use plain environment variables. I think that is the correct way to do it although I'm not familiar with systemd neither. @kakra I would appreciate if you can take a look just in case there is a better approach to this or I did something wrong.

I think I see why this warning is:

First, systemd devs advise to not use `Environment*=` at all. It has been considered a bad design decision. But to integrate best and easiest with how Gentoo handles runtime configuration (and preventing duplicate efforts), we need to do that.

Second, the `conf.d` files are actually bash code and could include code to generate configuration. This does not work with systemd, it has a simple parser and considers the file to only have static key/value assignments.

The only way around would be a wrapper which is executed before the service process using `ExecStartPre`: Such a script would parse the `conf.d` file using bash (via `source`) and export all variables. There are already packages in Gentoo which do it this way. One example is `elasticsearch`. I don't like that approach very much because it increases complexity of the service starter and eliminates the benefits which make systemd so fast.

But as long as this package doesn't distribute dynamic configuration in `conf.d`, I'd raise no concern. It's just something to keep in mind. Maybe put a hint in the `conf.d` file to not use code there if running with systemd. I also don't see a point here why this package could need dynamic code for configuration.

`bees` is another example for a package which needs to dynamically configure the environment before starting the service. But it brings it's own wrapper to do that. So that package is doing it right. Putting code in configuration files should be generally avoided.

**Comment by stkw0** [comment]
Thank you very much for the explanation!

**Pull request closed without merging**